README.md

```diff
@@ -1,34 +1,112 @@
+  - [Usage](#usage)
+    - [Inputs](#inputs)
+    - [Outputs](#outputs)
+  - [Examples](#examples)
+    - [Upload an Individual File](#upload-an-individual-file)
+    - [Upload an Entire Directory](#upload-an-entire-directory)
+    - [Upload using a Wildcard Pattern](#upload-using-a-wildcard-pattern)
+    - [Upload using Multiple Paths and Exclusions](#upload-using-multiple-paths-and-exclusions)
+    - [Altering compressions level (speed v. size)](#altering-compressions-level-speed-v-size)
+    - [Customization if no files are found](#customization-if-no-files-are-found)
+    - [(Not) Uploading to the same artifact](#not-uploading-to-the-same-artifact)
+    - [Environment Variables and Tilde Expansion](#environment-variables-and-tilde-expansion)
+    - [Retention Period](#retention-period)
+    - [Using Outputs](#using-outputs)
+      - [Example output between steps](#example-output-between-steps)
+      - [Example output between jobs](#example-output-between-jobs)
+  - [Limitations](#limitations)
+    - [Number of Artifacts](#number-of-artifacts)
+    - [Zip archives](#zip-archives)
+    - [Permission Loss](#permission-loss)
+  - [Where does the upload go?](#where-does-the-upload-go)
+
+
+## v4 - What's new
+
+> [!IMPORTANT]
+> upload-artifact@v4+ is not currently supported on GHES yet. If you are on GHES, you must use [v3](https://github.com/actions/upload-artifact/releases/tag/v3).
+
+The release of upload-artifact@v4 and download-artifact@v4 are major changes to the backend architecture of Artifacts. They have numerous performance and behavioral improvements.
+
+For more information, see the [`@actions/artifact`](https://github.com/actions/toolkit/tree/main/packages/artifact) documentation.
+
+### Improvements
+
+1. Uploads are significantly faster, upwards of 90% improvement in worst case scenarios.
+2. Once uploaded, an Artifact ID is returned and Artifacts are immediately available in the UI and [REST API](https://docs.github.com/en/rest/actions/artifacts). Previously, you would have to wait for the run to be completed before an ID was available or any APIs could be utilized.
+3. The contents of an Artifact are uploaded together into an _immutable_ archive. They cannot be altered by subsequent jobs. Both of these factors help reduce the possibility of accidentally corrupting Artifact files.
+4. The compression level of an Artifact can be manually tweaked for speed or size reduction.
+
+### Breaking Changes
 
-🚧 Under construction 🚧
+1. On self hosted runners, additional [firewall rules](https://github.com/actions/toolkit/tree/main/packages/artifact#breaking-changes) may be required.
+2. Uploading to the same named Artifact multiple times.
 
-Big changes coming...
+    Due to how Artifacts are created in this new version, it is no longer possible to upload to the same named Artifact multiple times. You must either split the uploads into multiple Artifacts with different names, or only upload once. Otherwise you _will_ encounter an error.
```

Comment:
So basically this breaks any workflow that were using this meaningful pattern?

In such cases you want to upload multiple assets from a matrix to a single artifact (https://github.com/docker/buildx/blob/8484fcdd5715c0683e8605e05cdffc0bfa8fde94/.github/workflows/build.yml#L149-L190) that you can then download from a dependent job to create a GitHub Release for example (https://github.com/docker/buildx/blob/8484fcdd5715c0683e8605e05cdffc0bfa8fde94/.github/workflows/build.yml#L244-L278).

How to do this then with v4?

Reply:
Make AI comments available to the members.